### ISSUE TITLE
v5 regression: type guard doesn't narrow type

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#53063]
# Bug Report

### 🕗 Version & Regression Information

- This changed between versions 4.9.5 and 5

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.0-dev.20230301#code/JYOwLgpgTgZghgYwgAgGJQig3gKGcgBwwFtgBXYgLmXgBsBnCAbhwF8cdRJZEUAFEuWLJc+IhFIVqYKGWZsOYAJ4EUAVRDAA9iGQBePMgA+Iw-jGCpNOA3nnk7fCdH3ClqshlyW91ixwAJhAItHAYyAg69GARABbBANYAkvQCEkLUABRw1BraIACU+gB8yHDIwPTIeTrIAGTIaZLE-pEg0TT6yJkAbjZyuZo6RXqlLsAw3QCECPEIyanuvf0QBUUu+AD0m8iAoOTIACoAysg9ACzUAAZ0jJdmyNvIgDLkhyc9AKxXAEZaWrQQcBAd3sfVocgAdOJmj5zI9zAA9AD8hnYfhwQA)

### 💻 Code

```ts
interface Free {
  premium: false;
}

interface Premium {
  premium: true;
}

type Union =
  | {
      premium: false;
    }
  | {
      premium: true;
    };

declare const checkIsPremium: (a: Union) => a is Union & Premium;

const f = (value: Union) => {
  if (!checkIsPremium(value)) {
    // ✅ TS v4: `false`
    // ❌ TS v5: `boolean`
    value.premium;
    //    ^?
  }
};
```

### 🙁 Actual behavior

See code comment above.

### 🙂 Expected behavior

See code comment above.

---

The issue disappears if we update `Union` like so:

```diff
 type Union =
   | {
       premium: false;
     }
-  | {
-      premium: true;
-    };
+  | Premium;
```

**Comment by Andarist** [comment]
This is already being worked on [here](https://github.com/microsoft/TypeScript/pull/52984)

**Comment by jakebailey** [comment]
Maybe this is a reduction of a larger case, but, why write the type guard here as:

```ts
declare const checkIsPremium: (a: Union) => a is Union & Premium;
```

Rather than this?

```ts
declare const checkIsPremium: (a: Union) => a is Premium;
```

**Comment by ahejlsberg** [comment]
Fixed by #52984.
